src/settings/target/mod.rs

```diff
@@ -266,88 +247,73 @@ impl Manifest {
-                            false // wrangler publish --release w/o workers_dev is zoned deploy
-                        }
-                    }
-                } else if let Some(wdd) = self.workers_dev {
-                    if wdd {
-                        if let Some(route) = &self.route {
-                            if !route.is_empty() {
-                                failure::bail!(wdd_failure)
-                            }
-                        }
-                    }
-                    wdd
-                } else {
-                    message::warn(&use_dot_dev_warning);
-                    true // wrangler publish w/o workers_dev is zoneless deploy
-                }
-            }
+    ) -> Result<Option<String>, failure::Error> {
+        let conflicting_targets_failure = "Your environment should only include `workers_dev` or `route`. If you are trying to publish to workers.dev, remove `route` from your wrangler.toml, if you are trying to publish to your own domain, remove `workers_dev`.";
+        let pick_target_failure =
+            "You must specify either `workers_dev` or `route` and `zone_id` in order to publish.";
+        let empty_route_failure =
+            "If you want to deploy to workers.dev, remove `route` from your environment config.";
+
+        log::debug!("top level workers_dev: {:?}", self.workers_dev);
+        log::debug!("top level route: {:?}", self.route);
+
+        // start with top level configuration
+        let (top_workers_dev, top_route) = match (self.workers_dev, self.route.clone()) {
+            (None, Some(route)) => (false, Some(route)),
+            (Some(workers_dev), None) => (workers_dev, None),
+            (Some(workers_dev), Some(route)) => (workers_dev, Some(route)),
+            (None, None) => (false, None),
+        };
 
-            // environment configuration
-            Some(environment) => {
-                if let Some(wdd) = environment.workers_dev {
-                    if wdd && environment.route.is_some() {
-                        failure::bail!(wdd_failure)
-                    }
-                    wdd
-                } else if let Some(wdd) = self.workers_dev {
-                    if wdd && environment.route.is_some() {
-                        false // allow route to override workers_dev = true if wdd is inherited
+        // override top level with environment
+        let (workers_dev, route) = if let Some(env) = &environment {
+            log::debug!("env workers_dev: {:?}", env.workers_dev);
+            log::debug!("env route: {:?}", env.route);
+            match (env.workers_dev, env.route.clone()) {
+                (None, Some(route)) => {
+                    if top_workers_dev && route.is_empty() {
+                        failure::bail!(empty_route_failure)
                     } else {
-                        wdd // inherit from top level
+                        (false, Some(route))
```

Comment:
so if there is a route in the environment, that always overrides whatever setting is in the top level configs. I'd LOVE to see a full, drawn out truth table for this in a comment before this code section, it would be great for maintainers!

Reply:
agreed! i can draw this out